hail/src/main/scala/is/hail/expr/ir/agg/AppendOnlyBTree.scala

```diff
@@ -234,52 +234,55 @@ class AppendOnlyBTree(kb: EmitClassBuilder[_], val key: BTreeKey, region: Value[
       }
     }
 
-    cb.invokeCode[Long](insertAt, nodec, insertIdxc, castKCode, childC)
+    cb.invokeCode[Long](insertAt, cb.this_, nodec, insertIdxc, castKCode, childC)
   }
 
-  private def getF(cb: EmitCodeBuilder, root: Value[Long], kc: EmitCode): Value[Long] = {
-    val get = kb.genEmitMethod("btree_get", FastSeq[ParamType](typeInfo[Long], kc.emitParamType), typeInfo[Long])
-    get.emitWithBuilder { cb =>
-      val node = get.getCodeParam[Long](1)
-      val k = get.getEmitParam(cb, 2)
-
-      val cmp = cb.newLocal("btree_get_cmp", -1)
-      val keyV = cb.newLocal("btree_get_keyV", 0L)
-
-      def insertOrGetAt(i: Int) = {
-        cb.if_(isLeaf(cb, node), {
-          cb.assign(keyV, insert(cb, node, const(i), k, const(0L)))
-          cb.assign(cmp, 0)
-        }, {
-          cb.assign(node, loadChild(cb, node, i - 1))
-        })
-      }
+  private def getF(param: EmitParamType): EmitMethodBuilder[_] =
+    kb.getOrGenEmitMethod("btree_get", ("btree_get", key),
```

Comment:
Because this method was being defined a lot of times.
I had completely changed the way emitclassbuilders stored methods and ran into lots of problems.
As part of debugging this, I saw that methods like these were being generated many many times.
Thus, caching the methods should probide us better jit behaviour.

Reply:
A better change would be for this class to take a module builder and define its own class I think. That would remove much of the complexity.

Reply:
Hmm, so you're saying you confirmed that the caching avoids generating multiple times in this case? I guess I'm wrong about
> As far as I can tell, using the BTreeKey as a key for the method hash will just use object identity, so this won't ever reuse an existing method

But I'm still a bit nervous about
> And it isn't entirely obvious to me that the same BTreeKey always guarantees interchangable get implementations

I'll think about that some more.